pkg/audit/workflow_auditor.go

```diff
@@ -0,0 +1,410 @@
+
+	event := NewAuditEvent(
+		EventTypeWorkflowStepCompleted,
+		source,
+		OutcomeSuccess,
+		subjects,
+		w.component,
+	)
+
+	target := map[string]string{
+		TargetKeyWorkflowID: workflowID,
+		TargetKeyStepID:     stepID,
+		TargetKeyType:       TargetTypeWorkflowStep,
+	}
+	event.WithTarget(target)
+
+	event.Metadata.Extra = map[string]any{
+		MetadataExtraKeyDuration:   duration.Milliseconds(),
+		MetadataExtraKeyRetryCount: retryCount,
+	}
+
+	event.LogTo(ctx, w.auditLogger, LevelAudit)
+}
+
+// LogStepFailed logs step failure.
+func (w *WorkflowAuditor) LogStepFailed(
+	ctx context.Context,
+	workflowID string,
+	stepID string,
+	duration time.Duration,
+	retryCount int,
+	_ error,
+) {
+	if !w.config.ShouldAuditEvent(EventTypeWorkflowStepFailed) {
+		return
+	}
+
+	source := w.extractSource(ctx)
+	subjects := w.extractSubjects(ctx)
+
+	event := NewAuditEvent(
+		EventTypeWorkflowStepFailed,
+		source,
+		OutcomeFailure,
+		subjects,
+		w.component,
+	)
+
+	target := map[string]string{
+		TargetKeyWorkflowID: workflowID,
+		TargetKeyStepID:     stepID,
+		TargetKeyType:       TargetTypeWorkflowStep,
+	}
+	event.WithTarget(target)
+
+	event.Metadata.Extra = map[string]any{
+		MetadataExtraKeyDuration:   duration.Milliseconds(),
+		MetadataExtraKeyRetryCount: retryCount,
+	}
+
+	event.LogTo(ctx, w.auditLogger, LevelAudit)
+}
+
+// LogStepSkipped logs conditional step skip.
+func (w *WorkflowAuditor) LogStepSkipped(
+	ctx context.Context,
+	workflowID string,
+	stepID string,
+	condition string,
+) {
+	if !w.config.ShouldAuditEvent(EventTypeWorkflowStepSkipped) {
+		return
+	}
+
+	source := w.extractSource(ctx)
+	subjects := w.extractSubjects(ctx)
+
+	event := NewAuditEvent(
+		EventTypeWorkflowStepSkipped,
+		source,
+		OutcomeSuccess,
+		subjects,
+		w.component,
+	)
+
+	target := map[string]string{
+		TargetKeyWorkflowID: workflowID,
+		TargetKeyStepID:     stepID,
+		TargetKeyType:       TargetTypeWorkflowStep,
+	}
+	event.WithTarget(target)
+
+	// Add condition as metadata
+	if condition != "" {
+		event.Metadata.Extra = map[string]any{
+			"condition": condition,
+		}
+	}
+
+	event.LogTo(ctx, w.auditLogger, LevelAudit)
+}
+
+// extractSource extracts source information from context.
+// For workflows, source is always local since they're internal orchestration.
+func (*WorkflowAuditor) extractSource(_ context.Context) EventSource {
+	return EventSource{
+		Type:  SourceTypeLocal,
+		Value: "vmcp-composer",
+		Extra: map[string]any{},
+	}
+}
+
+// extractSubjects extracts subject information from context.
+func (*WorkflowAuditor) extractSubjects(ctx context.Context) map[string]string {
+	subjects := make(map[string]string)
+
+	// Extract user information from Identity
+	if identity, ok := auth.IdentityFromContext(ctx); ok {
+		if identity.Subject != "" {
+			subjects[SubjectKeyUserID] = identity.Subject
+		}
+
+		if identity.Name != "" {
+			subjects[SubjectKeyUser] = identity.Name
+		} else if identity.Email != "" {
```

Comment:
Noticed the fallback order here is different from the HTTP auditor (`auditor.go:364-370`) - this one does Name → Email → PreferredUsername, while the HTTP auditor does Name → PreferredUsername → Email.

Might be worth extracting a shared helper function so they stay in sync? Also the HTTP version has some extra empty-string checks that this one doesn't have.